### ISSUE TITLE
Handle async actions

### ISSUE THREAD

**Issue by jgaskins** [clearwater-rb/grand_central#1]
Right now, all actions are handled synchronously. I've been thinking about ways to handle async actions for a while and @wied03 asked me about them, so I figured I'd open an issue to discuss them a bit more formally.

The idea I had was to add a special `AsyncAction` class so you could do something like this:

```ruby
# The AsyncAction would signal the store to add then/fail handlers
FetchThing = GrandCentral::AsyncAction.with_attributes(:id) do
  # The promise method is what the then/fail handlers would
  def promise
    Browser::HTTP.get("/api/things/#{id}")
  end
end

# Typical synchronous LoadX action: convert a hash (i.e. from JSON) to the right kind of object
LoadThing = GrandCentral::Action.with_attributes(:hash) do
  def thing
    Thing.new(hash)
  end
end

store.dispatch FetchThing.new(thing_id) do |response|
  store.dispatch LoadThing.new(response.json[:thing])
end
```

**Comment by wied03** [comment]
What about giving `dispatch` a promise of an action and doing it that way? Pros/cons?

**Comment by jgaskins** [comment]
@wied03 Do you mean dispatching a promise that yields an action on resolve?

```ruby
async_event = proc do
  promise = Promise.new

  Browser::HTTP.get(''/api/things/#{id}").then |response|
    promise.resolve(LoadThing.new(response))
  end

  promise
end

store.dispatch async_event.call
```

**Comment by wied03** [comment]
Maybe, it could also look like your prior example:

```ruby
# with_attributes will return a Promise of an action, which gets resolved
# when the promise returned by this block is resolved
FetchThing = GrandCentral::AsyncAction.with_attributes(:id) do
  Browser::HTTP.get("/api/things/#{id}")
end
```

Then:
```ruby
store.dispatch FetchThing.new(thing_id)
```

Store changes like this:
```ruby
def dispatch action
  # I forget whether Promise#value will detect if the arg a promise and just return
  action_promise = action.is_a?(Promise) ? action : Promise.value(action)
  action_promise.then do |action|
    old_state = state
    @state = @reducer.call state, action
    run_callbacks old_state, state
    self
  end      
end
```

**Comment by jgaskins** [comment]
Usually when I do this, what I actually want is to dispatch two actions: one to update the app state to signify that I'm loading that particular model and one to signify I'm no longer loading that model, turn the HTTP response into a model, and load that model into the app's state. One has to come before the HTTP request (and probably trigger it) and the other has to come after it (and possibly process it).

So while I like your solution, but it only solves one part. :-)

The whole use case I've got in mind is something like this: If a user navigates to a route where we don't have the model(s) loaded, we would want to display a "Loading…" message while we fetch the data, which means the store needs to know that we're fetching it. We wouldn't just be able to use `nil` or an empty array as sentinel values to indicate loading since they could be legitimate values for the model (if it doesn't exist) or collection (if it has no elements).

**Comment by wied03** [comment]
I debated whether to include that.. Seems like the way Redux does is if you give it a closure, it calls that with the `dispatch` function. I'm not sure exactly how that would work here. I think the equivalent might be:

```ruby
def dispatch action
  action = action[lambda { |*args| dispatch *args }] if action.is_a?(Proc)
  # I forget whether Promise#value will detect if the arg a promise and just return
  action_promise = action.is_a?(Promise) ? action : Promise.value(action)
  action_promise.then do |action|
    old_state = state
    @state = @reducer.call state, action
    run_callbacks old_state, state
    self
  end      
end

**Comment by jgaskins** [comment]
I'm trying to understand your idea from the perspective of the app dispatching these actions. Is this kinda how you're thinking it would be called?

```ruby
store.dispatch proc { |dispatch|
  fetch(url).then do |response|
    dispatch[LoadThing.new(response.json)]
  end
}
```

**Comment by wied03** [comment]
Yep. That's what I was thinking and if I'm right, that mirrors how Redux does it. I'm not saying that's the best way, more just food for thought.

**Comment by jgaskins** [comment]
Absolutely, I did open this issue specifically to invite discussion and I really do appreciate our conversations on topics like these. I don't mean to sound ungrateful. Even if I disagree, your input at least pushes me to figure out _why_ I disagree, which has been quite helpful. :-)

And definitely, dispatching a function which in turn is simply called totally makes sense for Redux because its use of middleware gives a pipeline to transform or even [terminate](https://github.com/gaearon/redux-thunk/blob/master/src/index.js) the dispatch and its use of action-creator functions allows them to be composed in a pretty idiomatic JS style. However, since we aren't using middleware or action-creator functions (our actions are typically instances of classes), I don't think it's a good fit here.

What I'm looking for, I think, is a way to say "this action somehow contains a promise that, when resolved, needs to dispatch this other action". So the first action's class would act as an "action creator" in the Redux sense, which knows how to generate the promise. Then whatever is doing the dispatching knows it can pass the block to indicate what should happen after. I'll need to get to a computer to write up a few more examples of what I'm thinking. Writing code on an iOS keyboard is gross. :-)

**Comment by wied03** [comment]
Something like this then?

```ruby
class AsyncAction
  attr_reader :other_action_promise, :promise

  def initialize(other_action_promise)
   @other_action_promise = other_action_promise
   @promise = yield
   end
  end
end

def dispatch action
  # I forget whether Promise#value will detect if the arg a promise and just return
  action_promise = action.is_a?(AsyncAction) ? action.promise : Promise.value(action)
  action_promise.then do |action|
    old_state = state
    @state = @reducer.call state, action
    run_callbacks old_state, state
    if action.is_a?(AsyncAction) && action.other_action_promise
      action.other_action_promise.then do |next_action|
        dispatch next_action
      end
    end
    self
  end      
end
```

**Comment by jgaskins** [comment]
What would the code look like that uses that?

**Comment by wied03** [comment]
I actually think this could look cleaner. I don't think a promise is needed for the 1st action (it will immediately come back in order to update the state to 'in progress' or what not). It's the subsequent ones. Maybe something like this then?

```ruby

# Do some sort of construction here
InProgressAction = AsyncAction.with_attributes
FinishedAction = AsyncAction.with_attributes(:result)

dispatch InProgressAction.new do
  Browser::HTTP.get("/api/things/#{id}").then do |result|
    FinishedAction.new(result)
  end
end

class AsyncAction
  attr_reader :next_promise

  def initialize
    result = yield
    @next_promise = result.is_a?(Promise) && result
  end
end

def dispatch action
  old_state = state
  @state = @reducer.call state, action
  run_callbacks old_state, state
  if action.is_a?(AsyncAction) && action.next_promise
    action.next_promise.then do |next_action|
      dispatch next_action
    end
  end
  self  
end
```

**Comment by jgaskins** [comment]
It seems like dispatching something like a `FetchThing` action _and_ specifying the HTTP request is a little redundant. I would imagine dispatching an action would also trigger its associated functionality.

Also, a problem that also exists in my original example: I just noticed that whether the block is for the `InProgressAction` constructor or the `dispatch` method is really ambiguous. In fact, in your usage example, the block would go to `dispatch`, but your `AsyncAction` constructor implies it should be there. That's definitely gonna confuse people. :-)

I wonder if `dispatch` should return the action instead of `self`. Then we could have `AsyncAction` delegate `then`, `fail`, and `always` to its promise.

```ruby
FetchThing = GrandCentral::AsyncAction.with_attributes(:id) do
  def promise
    @promise ||= fetch("/api/things/#{id}")
  end
end
LoadThing = GrandCentral::Action.with_attributes(:attributes)

store.dispatch(FetchThing.new(thing_id)).then do |response|
  store.dispatch LoadThing.new(response.json[:thing])
end.fail do |exception|
  warn "Could not load thing ##{thing_id}: #{exception.class} - #{exception.message}"
end
```

**Comment by jgaskins** [comment]
Hmm, I also just realized that having different classes for async vs sync actions makes using inheritance for action pattern matching harder. One of the things I like to do is split up action handlers by the parent type of action. It's similar to combining reducers with Redux:

```ruby
ThingsAction = Class.new(GrandCentral::Action)
DeleteThing = ThingsAction.with_attributes(:thing)

ThingAction = Class.new(ThingAction)
UpdateThing = ThingAction.with_attributes(:id)

# handle just actions pertaining to a particular thing
thing = proc do |current_thing, action|
  # ...
end

# handle actions pertaining to a collection of things
things = proc do |current_things, action|
  case action
  when ThingAction
    current_things.map { |t| thing[t, action] }
  when DeleteThing
    current_things - [action.thing]
  end
end

store = Store.new(initial_state) do |state, action|
  case action
  when ThingsAction
    state.merge(
      # Delegate handling of all ThingsActions to the things function
      things: things[state[:things], action],
    )
  else
    state
  end
end
```

The problem is that using different superclasses means a `FetchThings` action won't get grouped automatically with other `ThingsAction`s.

```ruby
ThingsAction = Class.new(GrandCentral::Action)
FetchThings = Class.new(GrandCentral::AsyncAction) # a whole different parent class
LoadThings = ThingsAction.with_attributes(:data)

ThingAction = Class.new(ThingsAction) # Specify actions for individual things
UpdateThing = ThingAction.with_attributes(:id, :new_attributes)
DeleteThing = ThingAction.with_attributes(:id)
FetchThing = GrandCentral::AsyncAction.with_attributes(:id) # a whole different parent class
LoadThing = ThingAction.with_attributes(:attributes)

# other handlers ...

store = GrandCentral::Store.new(initial_state) do |state, action|
  case action
  when ThingsAction
    things[state[:things], action]
  when FetchThings # Should be included in the above case
    # ...
  when FetchThing # Again ...
    # ...
  else
    state
  end
end
```

This isn't the worst thing, because we can do `when ThingsAction, FetchThings, FetchThing` to combine all three, but I would consider a `FetchThings` action a `ThingsAction`, so it should match automatically.

**Comment by wied03** [comment]
Maybe just use the same class for async vs. sync, using respond_to, and follow your idea to return the action (except return a promise if the action supplies one).

```ruby
FetchThing = GrandCentral::Action.with_attributes(:id) do
  def promise
    @promise ||= fetch("/api/things/#{id}")
  end
end
LoadThing = GrandCentral::Action.with_attributes(:attributes)
```

The usage of this is the same as your example:

```ruby
store.dispatch(FetchThing.new(thing_id)).then do |response|
  store.dispatch LoadThing.new(response.json[:thing])
end.fail do |exception|
  warn "Could not load thing ##{thing_id}: #{exception.class} - #{exception.message}"
end
```

Store looks like this:

```ruby
def dispatch action
  old_state = state
  @state = @reducer.call state, action
  run_callbacks old_state, state
  return action.promise if action.respond_to?(:promise)
  action  
end
```

**Comment by jgaskins** [comment]
What is your opinion of proxying the `then`, `fail`, and `always` calls on the action so we could just return the action without checking whether it responds to `promise`?

**Comment by wied03** [comment]
I suppose that's an OK option. If people used `is_a?(Promise)` in code, that wouldn't work but the reducer shouldn't care if the Redux principles of synchronous/side effect free code are followed (I think agree with those). The only other place left is the Store code, which of course is "under control" here.

**Comment by jgaskins** [comment]
Checking types with `return_value.is_a?(Promise)` is not something I care much about supporting. :-) Checking `return_value.respond_to?(:then)` is much more valuable here.

**Comment by wied03** [comment]
Yeah, I was just trying to think of any side effects and the above was a long of way of saying "none that matter"

**Comment by jgaskins** [comment]
Ah, okay. After another read-through of your comment, that makes sense. :-)

**Comment by jgaskins** [comment]
Fixed in d36384328c0490733e9d315d9681051cc4b6d66b and released in 0.2.0.

**Comment by wied03** [comment]
:+1:
